stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -1415,3 +1415,50 @@ function normalize(v::AbstractVector, p::Real = 2)
         return T[]
     end
 end
+
+# Iterator for the rows() and cols()
+struct MatrixDimIterator{T}
+    row::Bool
+    length::Int
+    M::Matrix{T}
+end
+
+function Base.iterate(it::MatrixDimIterator, state = 1)
+    if state > it.length
+        return nothing
+    elseif it.row == true
+        obj = @view it.M[state, :]
```

Comment:
This is type-unstable.  Both branches return views, yes, but views are specially parameterized by their indices to ensure fast accesses.  Here this branch returns a SubArray with indices `Tuple{Int, Colon}` and in the other branch the indices are `Tuple{Colon, Int}`.  It's especially important for the individual elements returned from the iterator to be inferred properly as if it's unstable then users will have to pay the cost of dynamic dispatch in every iteration of their loop.

You can test this interactively with `@code_warntype` or in the tests file with `@test @inferred(first(eachrow(A))) == …`.